app/Http/Controllers/InvoiceController.php

```diff
@@ -32,21 +36,61 @@ public function viewInvoiceEdit(Reservation $reservation) {
         ]);
     }
 
+    // Handlers
+
+    public function handleUpdateInvoice(Reservation $reservation, Request $request) {
+        $invoice = $reservation->invoice;
+
+        $this->validateUpdateInvoice($request);
+
+        $this->updateInvoice($invoice, $request);
+        return redirect(route('invoice.info', $reservation->id));
+    }
+
+    private function validateUpdateInvoice(Request $request) {
+        return $request->validate([
+            'payment_method' => 'required',
+            'value_added_tax' => 'required',
+            'description' => 'required',
+        ]);
+    }
+
+    public function handleRecalculateInvoice(Reservation $reservation, Request $request) {
+        $invoice = $reservation->invoice;
+
+        $this->validateUpdateInvoice($request);
+
+        $invoice->update([
+            'value_added_tax' => $request->value_added_tax,
+            'final_amount' => $this->calculateTotalInvoiceCosts($invoice, $request->value_added_tax),
+        ]);
+
+        return redirect(route('invoice.edit', $reservation->id));
+    }
+
+    // Methods
+
     public function createInvoiceForReservation(int $reservationId) {
         // TODO: fill in all values, remove use of factory
         Invoice::factory()->create([
             'reservation_id' => $reservationId,
         ]);
     }
 
-    public function handleUpdateInvoice(Reservation $reservation, Request $request) {
-        dd("Update functionality is not yet implemented");
-    }
+    private function updateInvoice(Invoice $invoice, Request $request) {
+        $totalCosts = $this->calculateTotalInvoiceCosts($invoice, $request->value_added_tax);
 
-    public function handleRecalculateInvoice(Reservation $reservation, Request $request) {
-        dd("Recalculate functionality is not yet implemented");
+        $invoice->update([
+            'value_added_tax' => $request->value_added_tax,
+            'payment_method_id' => $request->payment_method,
+            'description' => $request->description,
+            'is_paid' => isset($request->is_paid),
+            'final_amount' => $totalCosts,
+        ]);
     }
 
+    // Calculations
```

Comment:
Is calculation hier nodig?
Het staat al in de naam van de functies

Reply:
Ik vind het fijn om functies te categoriseren als er best veel in een bestand staan, dat is meer de reden daarachter